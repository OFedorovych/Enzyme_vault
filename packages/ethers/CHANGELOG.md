# @enzymefinance/ethers

## 4.0.5

### Patch Changes

- 8fe3d4a57: Publishing latest release

## 4.0.4

### Patch Changes

- 84c2e211e: Lock bn.js dependency version
- c030c92cd: Aave claimRewardsToSelf selector

## 4.0.3

### Patch Changes

- e8d978204: Upgrading ethers.js

## 4.0.2

### Patch Changes

- b4fcf8bc: Update packages after audit feedback and fixes.

## 4.0.1

### Patch Changes

- 2b54b39f: Add devtools to monorepo
