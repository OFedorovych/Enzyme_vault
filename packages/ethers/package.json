{
  "name": "@enzymefinance/ethers",
  "version": "4.0.5",
  "description": "Technology Regulated and Operated Investment Funds",
  "author": "Enzyme Council DAO <council@enzyme.finance>",
  "license": "MIT",
  "homepage": "https://enzyme.finance",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/avantgardefinance/ethereum-devtools.git"
  },
  "bugs": {
    "url": "https://github.com/avantgardefinance/ethereum-devtools/issues"
  },
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    }
  },
  "scripts": {
    "build": "tsup",
    "lint": "eslint \"src/**/*.{js,ts}\" --cache",
    "format": "pnpm lint --fix",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "@ethersproject/abi": "^5.6.4",
    "@ethersproject/abstract-signer": "^5.6.2",
    "ethers": "^5.6.8"
  },
  "devDependencies": {
    "eslint": "^8.12.0",
    "tsup": "^6.2.1",
    "typescript": "^4.5.5"
  },
  "tsup": {
    "entry": [
      "src/index.ts"
    ],
    "format": [
      "esm",
      "cjs"
    ],
    "dts": true,
    "splitting": true,
    "clean": true,
    "silent": true,
    "sourcemap": true
  },
  "eslintConfig": {
    "extends": "@enzymefinance/eslint-config",
    "rules": {
      "@typescript-eslint/no-shadow": "off"
    },
    "parserOptions": {
      "project": [
        "./tsconfig.json"
      ]
    }
  }
}
