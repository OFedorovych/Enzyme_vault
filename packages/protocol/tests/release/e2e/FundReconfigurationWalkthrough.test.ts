import type { ComptrollerLib, FundDeployer, VaultLib } from '@enzymefinance/protocol';
import {
  entranceRateBurnFeeConfigArgs,
  FeeManagerActionId,
  feeManagerConfigArgs,
  ITestStandardToken,
  managementFeeConfigArgs,
  managementFeeConvertRateToScaledPerSecondRate,
  ONE_DAY_IN_SECONDS,
  performanceFeeConfigArgs,
} from '@enzymefinance/protocol';
import type { ProtocolDeployment, SignerWithAddress } from '@enzymefinance/testutils';
import {
  addNewAssetsToFund,
  assertNoEvent,
  callOnExtension,
  createNewFund,
  createReconfigurationRequest,
  deployProtocolFixture,
  getAssetUnit,
} from '@enzymefinance/testutils';
import { BigNumber, utils } from 'ethers';

const FIVE_PERCENT = BigNumber.from(500);
const TEN_PERCENT = BigNumber.from(1000);

describe.each([['weth' as const], ['usdc' as const]])(
  'Walkthrough for %s as denomination asset',
  (denominationAssetId) => {
    let fork: ProtocolDeployment;
    let fundOwner: SignerWithAddress, investor: SignerWithAddress;

    let fundDeployer: FundDeployer;

    let comptrollerProxy: ComptrollerLib, vaultProxy: VaultLib;
    let feeManagerConfig: any;
    let denominationAsset: ITestStandardToken;

    beforeAll(async () => {
      fork = await deployProtocolFixture();

      [fundOwner, investor] = fork.accounts;

      fundDeployer = fork.deployment.fundDeployer;

      denominationAsset = new ITestStandardToken(
        denominationAssetId === 'weth' ? fork.config.weth : fork.config.primitives[denominationAssetId],
        provider,
      );

      // Create a fund with management and performance fee
      const managementFeeSettings = managementFeeConfigArgs({
        scaledPerSecondRate: managementFeeConvertRateToScaledPerSecondRate(utils.parseEther('0.01')),
      });
      const performanceFeeSettings = performanceFeeConfigArgs({
        rate: TEN_PERCENT,
      });
      const entranceRateBurnFeeSettings = entranceRateBurnFeeConfigArgs({ rate: FIVE_PERCENT });

      feeManagerConfig = feeManagerConfigArgs({
        fees: [fork.deployment.managementFee, fork.deployment.performanceFee, fork.deployment.entranceRateBurnFee],
        settings: [managementFeeSettings, performanceFeeSettings, entranceRateBurnFeeSettings],
      });

      // Buy shares in the fund for the fund owner
      const createFundTx = await createNewFund({
        denominationAsset,
        feeManagerConfig,
        fundDeployer,
        fundOwner,
        investment: {
          buyer: investor,
          provider,
          seedBuyer: true,
        },
        signer: fundOwner,
      });

      comptrollerProxy = createFundTx.comptrollerProxy;
      vaultProxy = createFundTx.vaultProxy;

      // Make both fees able to settle by adding free MLN (performance) and warping time (management)
      const mln = new ITestStandardToken(fork.config.primitives.mln, provider);
      const mlnUnit = await getAssetUnit(mln);

      await addNewAssetsToFund({
        provider,
        amounts: [mlnUnit],
        assets: [mln],
        comptrollerProxy,
        integrationManager: fork.deployment.integrationManager,
        signer: fundOwner,
      });
      await provider.send('evm_increaseTime', [ONE_DAY_IN_SECONDS * 30]);

      // Settle fees
      await callOnExtension({
        actionId: FeeManagerActionId.InvokeContinuousHook,
        comptrollerProxy,
        extension: fork.deployment.feeManager,
      });
    });

    it('signals a reconfiguration with the same setup', async () => {
      const { receipt } = await createReconfigurationRequest({
        denominationAsset,
        feeManagerConfigData: feeManagerConfig,
        fundDeployer,
        signer: fundOwner,
        vaultProxy,
      });

      expect(receipt).toMatchGasSnapshot(denominationAssetId);
    });

    // TODO: there are currently no fees that use "shares outstanding," otherwise we should test they are paid out
    it('warp beyond reconfiguration timelock and execute the reconfiguration', async () => {
      const reconfigurationTimelock = await fundDeployer.getReconfigurationTimelock();

      await provider.send('evm_increaseTime', [reconfigurationTimelock.toNumber()]);

      const receipt = await fundDeployer.connect(fundOwner).executeReconfiguration(vaultProxy);

      // Assert that DeactivateFeeManagerFailed did not fire
      assertNoEvent(receipt, comptrollerProxy.abi.getEvent('DeactivateFeeManagerFailed'));

      expect(receipt).toMatchGasSnapshot(denominationAssetId);
    });

    // TODO: finish this test suite with more stuff as-needed, it fills a specific need for now
    it.todo('more stuff');
  },
);
