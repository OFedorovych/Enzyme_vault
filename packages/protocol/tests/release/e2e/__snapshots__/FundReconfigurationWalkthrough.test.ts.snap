// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Walkthrough for usdc as denomination asset signals a reconfiguration with the same setup: usdc 1`] = `556810`;

exports[`Walkthrough for usdc as denomination asset warp beyond reconfiguration timelock and execute the reconfiguration: usdc 1`] = `352203`;

exports[`Walkthrough for weth as denomination asset signals a reconfiguration with the same setup: weth 1`] = `554552`;

exports[`Walkthrough for weth as denomination asset warp beyond reconfiguration timelock and execute the reconfiguration: weth 1`] = `319784`;
